### PR TITLE
Add forbidden-apis plugin; prohibit using system time zone

### PR DESCRIPTION
**Pull request by leventov**
This PR adds forbidden-apis plugin, which checks for usage of APIs which use the system locale and time zone. Also, a rule set added to prohibit using Joda-Time's APIs which use the system time zone.

This PR
 - Fixes a few remaining cases of usage of the system locale, after #4409.
 - Fixes usages of the system time zone in prod and test code.

Those 4K changed lines are largely in tests and reviewers could skip them.

In prod code, in many cases, this PR changes `new DateTime(timestamp)` to `DateTimes.utc(timestamp)`, but the created `DateTime` is effectively used to store/manipulate the underlying timestamp, without printing and parsing, so such changes don't have any practical effect.

While making this PR I identified several patterns of changes that have actual effect, if a Druid node is configured to have the system time zone which is not UTC:

 - Producing version string from the current `DateTime`, e. g. see `CustomVersioningPolicy`
 - Inserting current `DateTime` into metadata tables, and cases when current `DateTime` is used for various different purposes, e. g. see `IndexerSQLMetadataStorageCoordinator`, `SQLMetadataSupervisorManager`, `SQLMetadataRuleManager`, `SQLMetadataStorageUpdaterJobHandler`, `SQLMetadataSegmentPublisher`, `WorkerCuratorCoordinator`
 - Parsing user-provider DateTimes/Intervals, see `ClientInfoResource`
 - Cases when two DateTimes are created one after another, obviously meaning that they should be identical, but it's not guranteed. E. g. see `SimpleWorkerProvisioningStrategy`, `PendingTaskBasedWorkerProvisioningStrategy`, `TaskRunnerWorkItem`. I'm not sure I identified and fixed all such cases.
 - Changes in `druid-sql` module are special, I changed something but even not sure that it's right, @gianm could you please review those changes? Also, druid-sql uses GMT time zone on a few occasions instead of UTC, @gianm does that have a reason?

### CODE REVIEW

**Comment by leventov** [comment]
Also, at some places both in prod and test code, this PR changes constructs like `new DateTime().getMillis()` with `System.currentTimeMillis()`, should be OK but please review.

**Comment by gianm** [comment]
> Changes in druid-sql module are special, I changed something but even not sure that it's right, @gianm could you please review those changes? Also, druid-sql uses GMT time zone on a few occasions instead of UTC, @gianm does that have a reason?

The druid-sql changes look ok to me. The code using GMT is doing that because it's creating an object that Calcite will use internally, and Calcite expects it to be in GMT (it will throw an exception if not).

**Comment by leventov** [comment]
Probably fixes #2619

File: extensions-core/histogram/src/main/java/io/druid/query/aggregation/histogram/sql/QuantileSqlAggregator.java [hunk 1]
```diff
@@ -183,7 +183,7 @@ public Aggregation toDruidAggregation(
       }
     } else {
       final ExpressionVirtualColumn virtualColumn = input.toVirtualColumn(
-          String.format("%s:v", name),
+          StringUtils.format("%s:v", name),
```
Comment (drcrallen) [inline extensions-core/histogram/src/main/java/io/druid/query/aggregation/histogram/sql/QuantileSqlAggregator.java]: not date-time

File: hll/src/test/java/io/druid/hll/HyperLogLogCollectorTest.java [hunk 2]
```diff
@@ -118,15 +119,10 @@ public void testHighCardinalityRollingFold() throws Exception
...
-    System.out.println("True cardinality " + n);
-    System.out.println("Rolling buffer cardinality " + rolling.estimateCardinality());
-    System.out.println("Simple  buffer cardinality " + simple.estimateCardinality());
-    System.out.println(
-        StringUtils.format(
-            "Rolling cardinality estimate off by %4.1f%%",
-            100 * (1 - rolling.estimateCardinality() / n)
-        )
-    );
+    log.info("True cardinality " + n);
```
Comment (drcrallen) [inline hll/src/test/java/io/druid/hll/HyperLogLogCollectorTest.java]: not datetime

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/SimpleWorkerProvisioningStrategy.java [hunk 3]
```diff
@@ -103,8 +104,8 @@ public Provisioner makeProvisioner(WorkerTaskRunner runner)
     private final Set<String> currentlyTerminating = Sets.newHashSet();
 
     private int targetWorkerCount = -1;
-    private DateTime lastProvisionTime = new DateTime();
-    private DateTime lastTerminateTime = new DateTime();
+    private DateTime lastProvisionTime = DateTimes.nowUtc();
+    private DateTime lastTerminateTime = lastProvisionTime;
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/SimpleWorkerProvisioningStrategy.java]: why this change?

File: integration-tests/src/main/java/org/testng/TestNG.java [hunk 4]
```diff
@@ -329,17 +329,8 @@ public void initializeSuitesAndJarFile()
...
-          }
-          catch (ParserConfigurationException e) {
-            e.printStackTrace(System.out);
-          }
-          catch (SAXException e) {
-            e.printStackTrace(System.out);
-          }
-          catch (IOException e) {
-            e.printStackTrace(System.out);
+          catch (ParserConfigurationException | SAXException | IOException e) {
```
Comment (drcrallen) [inline integration-tests/src/main/java/org/testng/TestNG.java]: unrelated

File: java-util/src/main/java/io/druid/java/util/common/granularity/NoneGranularity.java [hunk 5]
```diff
@@ -42,23 +42,19 @@ public DateTimeFormatter getFormatter(Formatter type)
...
 
   @Override
   public DateTime bucketStart(DateTime time)
   {
-    if (time == null) {
-      return null;
-    }
-
-    return new DateTime(time.getMillis());
+    return time;
```
Comment (drcrallen) [inline java-util/src/main/java/io/druid/java/util/common/granularity/NoneGranularity.java]: There is no guarantee that the chronology of `time` is equal to the return contract, right? should this method do any sort of "utc" enforcement?

File: java-util/src/test/java/io/druid/java/util/common/parsers/TimestampParserTest.java [hunk 6]
```diff
@@ -38,125 +40,97 @@ public void testStripQuotes() throws Exception
...
     Assert.assertEquals(millis2, parser.apply("Wed Nov 9 04:00:00 UTC-0600 CST 1994").getMillis());
 
-    parser = ParserUtils.createTimestampParser("EEE MMM dd HH:mm:ss zZ yyyy");
+    parser = TimestampParser.createTimestampParser("EEE MMM dd HH:mm:ss zZ yyyy");
     Assert.assertEquals(millis1, parser.apply("Wed Nov 9 04:00:00 GMT-0800 1994").getMillis());
     Assert.assertEquals(millis2, parser.apply("Wed Nov 9 04:00:00 GMT-0600 1994").getMillis());
     Assert.assertEquals(millis1, parser.apply("Wed Nov 9 04:00:00 UTC-0800 1994").getMillis());
     Assert.assertEquals(millis2, parser.apply("Wed Nov 9 04:00:00 UTC-0600 1994").getMillis());
-
-    parser = ParserUtils.createTimestampParser("EEE MMM dd HH:mm:ss zZ Q yyyy");
```
Comment (drcrallen) [inline java-util/src/test/java/io/druid/java/util/common/parsers/TimestampParserTest.java]: why is this removed?

File: pom.xml [hunk 7]
```diff
@@ -58,6 +58,8 @@
     </scm>
 
     <properties>
+        <maven.compiler.target>1.8</maven.compiler.target>
```
Comment (drcrallen) [inline pom.xml]: how is this different than the prior setting?

File: processing/src/main/java/io/druid/query/expression/ExprUtils.java [hunk 8]
```diff
@@ -56,20 +58,21 @@ public static PeriodGranularity toPeriodGranularity(
...
-    }
-
     if (timeZoneArg == null) {
       timeZone = null;
     } else {
       final String value = timeZoneArg.eval(bindings).asString();
       timeZone = value != null ? DateTimeZone.forID(value) : null;
     }
 
+    Chronology chronology = timeZone == null ? ISOChronology.getInstanceUTC() : ISOChronology.getInstance(timeZone);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/expression/ExprUtils.java]: Chronology is only needed in the `else` case, right? can it be limited in scope to just that code block?

File: processing/src/main/java/io/druid/query/expression/TimestampExtractExprMacro.java [hunk 9]
```diff
@@ -70,7 +71,7 @@ public Expr apply(final List<Expr> args)
     }
 
     final Expr arg = args.get(0);
-    final Unit unit = Unit.valueOf(((String) args.get(1).getLiteralValue()).toUpperCase());
+    final Unit unit = Unit.valueOf(StringUtils.toUpperCase((String) args.get(1).getLiteralValue()));
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/expression/TimestampExtractExprMacro.java]: not datetime

File: processing/src/main/java/io/druid/query/metadata/SegmentAnalyzer.java [hunk 10]
```diff
@@ -249,8 +250,8 @@ private ColumnAnalysis analyzeStringColumn(
     }
 
     if (analyzingSize()) {
-      final long start = storageAdapter.getMinTime().getMillis();
-      final long end = storageAdapter.getMaxTime().getMillis();
+      final DateTime start = storageAdapter.getMinTime();
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/metadata/SegmentAnalyzer.java]: why the change away from `long`?

File: processing/src/main/java/io/druid/query/spec/LegacySegmentSpec.java [hunk 11]
```diff
@@ -49,17 +49,10 @@
...
-          public Interval apply(Object input)
-          {
-            return new Interval(input);
-          }
-        }
-    );
+    return intervalStringList
+        .stream()
+        .map(input -> new Interval(input, ISOChronology.getInstanceUTC()))
+        .collect(Collectors.toList());
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/spec/LegacySegmentSpec.java]: This is functionally different. The prior method had wrappers on fetching the underlying values, this materializes a new copy of the list, but it should be ok for here since you are discarding the prior list anyways. I don't think this is called in any tight loops, so should be safe to cause a bit of extra heap garbage.

File: processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQuery.java [hunk 12]
```diff
@@ -34,21 +35,16 @@
...
 
 /**
  */
 public class TimeBoundaryQuery extends BaseQuery<Result<TimeBoundaryResultValue>>
 {
-  public static final Interval MY_Y2K_INTERVAL = new Interval(
-      new DateTime("0000-01-01"),
-      new DateTime("3000-01-01")
-  );
+  private static final QuerySegmentSpec DEFAULT_SEGMENT_SPEC = new MultipleIntervalSegmentSpec(Intervals.ONLY_ETERNITY);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQuery.java]: this is used a lot, would it make sense to have it a protected static field in `BaseQuery`?

File: processing/src/main/java/io/druid/segment/data/BlockLayoutIndexedDoubleSupplier.java [hunk 13]
```diff
@@ -105,7 +106,7 @@ public void fill(int index, double[] toFill)
     {
       if (totalSize - index < toFill.length) {
         throw new IndexOutOfBoundsException(
-            String.format(
+            StringUtils.format(
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/data/BlockLayoutIndexedDoubleSupplier.java]: not datetime

File: processing/src/main/java/io/druid/segment/data/EntireLayoutIndexedDoubleSupplier.java [hunk 14]
```diff
@@ -61,7 +62,7 @@ public void fill(int index, double[] toFill)
     {
       if (totalSize - index < toFill.length) {
         throw new IndexOutOfBoundsException(
-            String.format(
+            StringUtils.format(
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/data/EntireLayoutIndexedDoubleSupplier.java]: not datetime

File: server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java [hunk 15]
```diff
@@ -316,7 +316,8 @@ public SegmentPublishResult announceHistoricalSegments(
     // Find which segments are used (i.e. not overshadowed).
     final Set<DataSegment> usedSegments = Sets.newHashSet();
     for (TimelineObjectHolder<String, DataSegment> holder : VersionedIntervalTimeline.forSegments(segments)
-                                                                                     .lookupWithIncompletePartitions(JodaUtils.ETERNITY)) {
+                                                                                     .lookupWithIncompletePartitions(
+                                                                                         Intervals.ETERNITY)) {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java]: formatting here looks weird?

File: server/src/main/java/io/druid/server/http/HostAndPortWithScheme.java [hunk 16]
```diff
@@ -56,8 +57,9 @@ public static HostAndPortWithScheme fromString(String scheme, String hostPortStr
 
   private static String checkAndGetScheme(String scheme)
   {
-    Preconditions.checkState(scheme.toLowerCase().equals("http") || scheme.toLowerCase().equals("https"));
-    return scheme.toLowerCase();
+    String schemeLowerCase = StringUtils.toLowerCase(scheme);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/server/http/HostAndPortWithScheme.java]: not datetime

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/SimpleWorkerProvisioningStrategy.java (hunk 3 above)
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/SimpleWorkerProvisioningStrategy.java]: When "now" DateTime objects are created one after another, they almost always will be identical, but sometimes, very rarely, they could differ. It creates a perfect ground for bugs and transient failures, e. g. if those date times are compared for equality, it may always pass tests, but rarely fail in production. In this class it is not yet important, because `lastProvisionTime` and `lastTerminateTime` are not compared in any way.

File: integration-tests/src/main/java/org/testng/TestNG.java (hunk 4 above)
Comment (leventov) [inline integration-tests/src/main/java/org/testng/TestNG.java]: `Exception.printStackTrace()` is also a thing that is forbidden in forbidden-apis, so had to make this change to make the project compile

File: extensions-core/histogram/src/main/java/io/druid/query/aggregation/histogram/sql/QuantileSqlAggregator.java (hunk 1 above)
Comment (leventov) [inline extensions-core/histogram/src/main/java/io/druid/query/aggregation/histogram/sql/QuantileSqlAggregator.java]: Forbidden-apis forbids uses of the default Locale, they were eliminated mostly in #4409, but some occurrences could have been missed, and some new were introduced since that PR was merged. This is mentioned in the PR description.

File: java-util/src/test/java/io/druid/java/util/common/parsers/TimestampParserTest.java (hunk 6 above)
Comment (leventov) [inline java-util/src/test/java/io/druid/java/util/common/parsers/TimestampParserTest.java]: This test was commented (for very long time, since 2012 or 2013): https://github.com/druid-io/druid/pull/4611/files/586d99cdea0e87f24a469ab681c167493d976a9a#diff-a7af8b2daf1212edc1fe51e80dc0b4dfL71 I uncommented it, but it was partially failing, and I removed the failing part. I think it was commented just because the test was failing. So this change at least doesn't reduce the test coverage.

File: java-util/src/main/java/io/druid/java/util/common/granularity/NoneGranularity.java (hunk 5 above)
Comment (leventov) [inline java-util/src/main/java/io/druid/java/util/common/granularity/NoneGranularity.java]: I don't think it should, I think it's anticipated that `NoneGranularity.bucketStart(time)` just returns the argument back without modification.

File: processing/src/main/java/io/druid/query/metadata/SegmentAnalyzer.java (hunk 10 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/metadata/SegmentAnalyzer.java]: Those `start` and `end` variables are used in an `Interval` constructor below. If it accepts `long` (timestamps), it uses the default system time zone. If it accepts DateTime objects, it borrows the timeZone from the `start`.

File: processing/src/main/java/io/druid/query/expression/ExprUtils.java (hunk 8 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/expression/ExprUtils.java]: Changed

File: processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQuery.java (hunk 12 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/timeboundary/TimeBoundaryQuery.java]: I don't want to expose it, because this spec is harmful if used for topN, groupBy, timeseries queries. So I limit it's scope to "special" query types.

File: server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java (hunk 15 above)
Comment (leventov) [inline server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java]: Fixed

**Comment by gianm** [comment]
I reviewed the druid-sql part and it looks good to me. I am also +1 on the general design. But I haven't done code level review of anything outside druid-sql.

Has anyone else reviewed the rest? Or does it still need code level review?

**Comment by leventov** [comment]
I think it needs code-level approve from @drcrallen (excluding tests)

**Comment by drcrallen** [comment]
back from vacation. checking again.

**Comment by leventov** [comment]
@drcrallen thanks for review. Could you please merge this PR?
